Add spiffe secrets engine docs (#1218)
hashicorp/web-unified-docs: content/vault/v2.0.x (rc)/content/api-docs/secret/spiffe.mdx

---
layout: api
page_title: SPIFFE - Secrets engines - HTTP API
description: Use the SPIFFE plugin API to mint SPIFFE JWT-SVIDs
---

# SPIFFE secrets engine (API)

Use the SPIFFE plugin to mint SPIFFE JWT-SVIDs. The example code below assumes
you mounted the SPIFFE plugin at `/spiffe`. Since it is possible to enable
secrets engines at any location, please update your API calls accordingly.

## Create configuration

Configure the SPIFFE trust domain for the plugin.

| Method | Path                  |
|:-------|:----------------------|
| `POST` | `/spiffe/config` |

### Request parameters

- `trust_domain` `(string: <required>)` - The SPIFFE trust domain used by the
  plugin. You can omit the `spiffe://` prefix.
- `bundle_refresh_hint` `(int or time string: "1h")` - The value to put as the refresh hint in trust bundles we publish.
May not exceed key_lifetime/10.
- `key_lifetime` `(int or time string: "24h")` - How often to generate a new signing key. Uses duration format strings.
- `jwt_issuer_url` `(string: optional)` - The base URL to use for JWT issuer claims (`iss`), including
  `https://` schema, host, and optional port. Must be reachable by whatever systems consume the JWTs.

>   By default it will be the vault cluster's api_addr.  The mount path and issuer endpoint will be appended to this value.

Reviewer: ```suggestion
  By default, Vault uses the cluster API address as the issuer URL. Vault
  appends the mount path and issuer endpoint to the issuer URL.
```
Style correction: write in active voice, avoid "this" as a pronoun

> - `jwt_signing_algorithm` `(string: "RS256")` - Signing algorithm to use. Allowed values are: RS256 (default),

Reviewer: ```suggestion
- `jwt_signing_algorithm` `(string: "RS256")` - Signing algorithm to use. Must be one of: RS256 (default),
```
Style correction: write in active voice

## Read configuration

Fetch the SPIFFE configuration details.

| Method | Path                  |
|:-------|:----------------------|
| `GET`  | `/spiffe/config` |

### Request parameters

None.

### Sample response

```json
{
  "bundle_refresh_hint": "3600",
  "jwt_issuer_url": "",
  "jwt_oidc_compability_mode": false,
  "jwt_signing_algorithm": "RS256",
  "key_lifetime": "86400",
  "trust_domain": "example.org"
}
```

## Create role

Create or update a named role that specifies the content of JWTs.

| Method | Path                      |
|:-------|:--------------------------|
| `POST` | `/spiffe/role/:name` |

### Path parameters

- `name` `(string: <required>)` - The unique name of the SPIFFE role you want to
  update.

### Request parameters

- `template` `(string: <required>)` - The template string to use for generating tokens. This may be in
  stringified JSON or base64 format.
- `ttl` `(int or time string: "5m")` - TTL of the tokens generated against the role. Uses duration format strings.
  TTLs of generated tokens will be capped by the lifespan of the current signing key.
- `use_jti_claim` `(bool: false)` - If true, the jti claim will be included in JWTs, which may impact

>   their reusability.

Reviewer: ```suggestion
- `use_jti_claim` `(bool: false)` - If true, the Vault includes the `jti` claim
  in JWTs, which may impact their reusability.
```
Style correction: write in active voice

Roles define a template in much the same way as [identity tokens](../identity/identity-token).  As with identity tokens, some claims cannot be
overridden by the template and are populated by the secrets engine:

- `iss`: Issuer URL is populated based on the global config setting jwt_issuer_url
- `aud`: Audience is populated by the JWT minting endpoint argument
- `iat`: Time of issue
- `exp`: Expiration time for the token (iat + ttl)
- `jti`: UUID that helps for audit trails and preventing replay attacks (rfc7519), present only if the role setting
         `use_jti_claim` is true

In addition to these standard claims, the following claim will be included to assist in provenance:
- `vault: {entity: {id: ENTITY_ID}}`
Here the entity id is that of the Vault client.

Unlike with identity tokens, `sub` can (and must) be specified.  This is either the SPIFFEID of the JWT SVID, or

> the path part of a SPIFFEID, in which case the secrets engine will prepend `spiffe://` and the configured trust domain.

Reviewer: ```suggestion
the path part of a SPIFFEID. The SPIFFE plugin automatically prepends path values with `spiffe://` and the configured trust domain.
```
Style correction: write in active voice

Role creation will fail if the parsed template doesn't have a sub claim.  SVID minting will fail if the expanded template

> isn't a valid SPIFFEID, or doesn't have the configured trust domain.

Reviewer: ```suggestion
Role creation fails if the parsed template does not have a sub claim. SVID
minting fails if the expanded template is not a valid SPIFFEID, or does not have
the configured trust domain.
```
Style correction: write in active voice, avoid contractions
